File: 백준/Silver/1940. 주몽/주몽.py
import sys
read = sys.stdin.readline

n = int(read())
m = int(read())
materials = list(map(int, read().split()))
materials.sort()

start_idx = 0
end_idx = n - 1
count = 0

while start_idx != end_idx:
    s = materials[start_idx] + materials[end_idx]
    if s == m:
        count += 1
        start_idx += 1
    elif s > m:
        end_idx -= 1
    else:
        start_idx += 1

print(count)

File: 백준/Silver/1940. 주몽/주몽.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n = Integer.parseInt(br.readLine());
		int m = Integer.parseInt(br.readLine());
		
		int[] materials = new int[n];
		
		StringTokenizer stk = new StringTokenizer(br.readLine());
		for(int i = 0; i < n; i++)
		{
			materials[i] = Integer.parseInt(stk.nextToken()); 
		}
		
		Arrays.sort(materials);
		
		int count = 0;
		int start = 0;
		int end = n - 1;
		
		while(start < end) {
			int sum = materials[start] + materials[end];
			
			if(sum > m) {
				end --;
			}
			else if(sum < m) {
				start++;
			}
			else {
				count++;
				start++;
				end--;
			}
		}
		System.out.println(count);
	}
}